R/LinkedMsExperiment.R

```diff
@@ -0,0 +1,366 @@
+#'
+#' @importMethodsFrom methods slot slot<-
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+#'
+#' @examples
+#'
+#' ## Add a new spectra variable to a Spectra
+#' .set_element(mse, "spectra.new.value", 1:3)
+#'
+#' ## Add a new metadata element
+#' .set_element(mse, "metadata.new_entry", data.frame(1:3))
+.set_element <- function(x, name = character(), value = NULL) {
+    if (length(name) == 0)
+        return(x)
+    name <- unlist(strsplit(name[1L], split = ".", fixed = TRUE))
+    slt <- name[1L]
+    if (!any(slotNames(x) == slt))
+        stop("No slot named '", slt, "' available in 'x'")
+    if (length(name) == 1)
+        slot(x, slt, check = FALSE) <- value
+    else {
+        el <- paste0(name[-1L], collapse = ".")
+        if (length(dim(slot(x, slt))))
+            slot(x, slt, check = FALSE)[, el] <- value
+        else
+            slot(x, slt) <- do.call("$<-", list(slot(x, slt), el, value))
+    }
+    x
+}
+
+#' @param x `character` similar to a join statement, e.g.
+#'     `"sampleData.mzML = spectra.dataOrigin"`
+#'
+#' @return `character` with slot from, column from, slot to, column to.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+#'
+#' @examples
+#'
+#' x <- "sampleData.mzml_file = spectra.dataOrigin"
+#' .parse_join_string(x)
+#'
+#' .parse_join_string("sampleData.mzml.file= spectra.dataOrigin")
+.parse_join_string <- function(x) {
+    parts <- unlist(strsplit(gsub(" ", "", x, fixed = TRUE), "="))
+    if (length(parts) != 2)
+        stop("unsupported format for the link description. ",
+             "Please see ?linkSampleData")
+    a <- unlist(strsplit(parts[1L], ".", fixed = TRUE))
+    b <- unlist(strsplit(parts[2L], ".", fixed = TRUE))
+    if (length(a) < 2 | length(b) < 2)
+        stop("unsupported format for the link desciption. ",
+             "Please see ?linkSampleData")
+    c(a[1L], paste(a[-1L], collapse = "."),
+      b[1L], paste0(b[-1L], collapse = "."))
+}
+
+#' Create a link matrix for specified vectors.
+#'
+#' @return `integer` `matrix` with the indices of the matches.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+#'
+#' @examples
+#'
+#' .link_matrix(1:3, 2:20)
+#' .link_matrix(c("a", "a", "b", "d", "b", "c"), c("g", "a", "b", "e"))
+.link_matrix <- function(from = integer(), to = integer()) {
+    res <- lapply(from, function(x) which(to == x))
+    ls <- lengths(res)
+    cbind(rep(seq_along(from), ls), unlist(res[ls > 0], use.names = FALSE))
+}
+
+#' @rdname LinkedMsExperiment
+setGeneric("linkSampleData", function(object, ...)
+    standardGeneric("linkSampleData"))
+#' @rdname LinkedMsExperiment
+setMethod("linkSampleData", "MsExperiment",
+          function(object, with = character(),
+                   fromIndex = seq_len(nrow(sampleData(object))),
+                   toIndex = integer()) {
+              object <- as(object, "LinkedMsExperiment")
+              if (!length(with))
+                  return(object)
```

Comment:
Here, you return a `LinkedMsExperiment` even though no links have been added. Wouldn't it be safer to return the original `object` in such case?

Reply:
I'm not sure. I think it would be confusing if `linkSampleData` returns in some instances a `MsExperiment` and in others `LinkedMsExperiment`. Returning a `LinkedMsExperiment` (even without links) would be more consistent.

Reply:
All functionality is now implemented for `MsExperiment`, thus this should be fixed too.